working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -115,240 +145,165 @@ Remember, the tests generated during the Copilot run in this lesson may not matc
     assert response_body == {"message": "book cat invalid"}
+```
 
-def test_update_book_nonexistent_id(client, two_saved_books):
-    # Arrange
-    book_id = 3
-    data = {
-        "title": "Updated Book",
-        "description": "New description"
-    }
+We already have a good range of scenarios:
+- successfully update a book
+- return an error if the book id is invalid
+- return an error if a book with the id doesn't exist
+
+Something to note is that the `test_update_book` test is using our `GET` `books/<book_id>` route to verify that the database changed. Instead of relying on another route, we could check the database itself. We can ask Copilot to make this change with a prompt like:
+
+> Please update these tests to verify changes using the database rather than client.get
 
+*Updated `test_update_book`:*
+```py
+def test_update_book(client, two_saved_books, db_session):
     # Act
-    response = client.put(f"/books/{book_id}", json=data)
+    response = client.put("/books/1", json={
+        "title": "Updated Ocean Book",
+        "description": "Updated description"
+    })
     response_body = response.get_json()
 
     # Assert
-    assert response.status_code == 404
-    assert response_body == {"message": "book 3 not found"}
+    assert response.status_code == 204
+    assert response_body is None
+
+    # Verify the update using the database
+    updated_book = db_session.query(Book).get(1)
+    assert updated_book.title == "Updated Ocean Book"
+    assert updated_book.description == "Updated description"
 ```
 
-This time Copilot did a great job writing tests that properly use our test fixtures and assert on the correct values! The tests above will run as is, and we already have a good range of scenarios:
-- successfully update a book
-- return an error if the book id is invalid
-- return an error if a book with the id doesn't exist
+We are using the database now, but from these new changes we need to address that:
+- we must import the `Book` class
+- the test is using a fixture `db_session` we haven't created
+- Copilot chose deprecated querying syntax `db_session.query(Book).get(1)`. 
+
+The overall frame of the tests is looking good so we can press "Accept" and save the file. We could see about writing a fixture `db_session` like Copilot suggested, but we will choose to import `db` into the test file since we need to use both the `select` method and `session` attribute from `db` to query for a specific record. 
+
+If we import `Book` and `db` and change our querying syntax over to using `db.select` and `db.session` our invalid and deprecated syntax issues are solved, but if we try to run the tests, we'll see a `JSONDecodeError`. This is because the line `response.get_json()` will raise a `JSONDecodeError` when the `response` object has an empty body. We'll address this last issue by either manually or with Copilot's help replacing the `response_body` assert with one that checks if `response.content_length is None`. 
```

Comment:
I think these commas help read more smoothly.
```suggestion
If we import `Book` and `db` and change our querying syntax over to using `db.select` and `db.session`, our invalid and deprecated syntax issues are solved, but if we try to run the tests, we'll see a `JSONDecodeError`. This is because the line `response.get_json()` will raise a `JSONDecodeError` when the `response` object has an empty body. We'll address this last issue by, either manually or with Copilot's help, replacing the `response_body` assert with one that checks if `response.content_length is None`. 
```